data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,670 @@
+//! Const-string related implementation and structs.
+extern crate alloc;
+
+#[cfg(not(feature = "std"))]
+use alloc::{boxed::Box, str::from_utf8_unchecked, string::String};
+use core::{
+    alloc::Layout,
+    cmp::{Eq, Ord, Ordering, PartialEq, PartialOrd},
+    convert::TryFrom,
+    hash::{Hash, Hasher},
+    mem::{size_of, ManuallyDrop},
+    ops::Deref,
+    slice::from_raw_parts,
+};
+#[cfg(feature = "std")]
+use std::str::from_utf8_unchecked;
+
+use derive_more::{AsRef, DebugCustom, Display, From};
+use iroha_schema::{IntoSchema, MetaMap};
+use parity_scale_codec::WrapperTypeEncode;
+use serde::{Serialize, Serializer};
+
+const MAX_INLINED_STRING_LEN: usize = 2 * size_of::<usize>() - 1;
+
+/// Immutable inlinable string.
+/// Strings shorter than 15/7/3 bytes (in 64/32/16-bit architecture) are inlined.
+#[derive(Clone, Display, DebugCustom, From, AsRef)]
+#[from(forward)]
+#[as_ref(forward)]
+#[display(fmt = "{}", "&**self")]
+#[debug(fmt = "{:?}", "&**self")]
+pub struct ConstString(ConstStringData);
```

Comment:
is this wrapper struct necessary? Wouldn't it be possible to implement everything directly on `ConstStringData`?